### DISCUSSION TITLE
A New Real Estate plugin is Here!

### DISCUSSION THREAD

**Discussion by Vodhin** [e107inc/e107#5238] [Show and tell]
This is a plugin for the e107 Content management System to manage Real Estate Agencies, Agents, and Listings. This plugin started as a way for an individual Real Estate agent to easily list their properties for sale and has evolved into one that can handle multiple agents and agencies. 

[**NEW version 1.1.0 is available**](https://github.com/Vodhin/estate/releases/tag/v1.1.0)
https://github.com/Vodhin/estate/releases/tag/v1.1.0

November 27 2024: So many changes, so many fixes. Added new functions to handle PHP based Currency and Date formatting based on PHP Locale settings which can be localized for each property listing. Currency formatting can also be forced to a spcific symbol before or after the number, or omitted altogether. This is handled by a new form that replaces the old Original Price and Listing Price fields (they are still present, but hidden). Included in the new form is also a new Price History table to record changes to the listing's price or status. Added ability to save custom Listing templates for each property: these can be altered either in the Admin area or directly on a Listing's View page.

Here is some eye-candy:
![instr000](https://github.com/e107inc/e107/assets/135135923/1b171130-9d25-46d3-9c4f-cfb3bba28f8f)

![instr013b](https://github.com/e107inc/e107/assets/135135923/9ce87fcd-f6b9-44a9-a30f-67630e91b063)

![instr_agtagencylist](https://github.com/e107inc/e107/assets/135135923/92910915-1b3f-409f-87bf-fd9f2d2a0f78)

![instr_agtagency](https://github.com/e107inc/e107/assets/135135923/65262b19-567e-48f3-ac71-7c167ba7d09e)

![instr_agtuserlist](https://github.com/e107inc/e107/assets/135135923/d1bec325-d315-411c-9d9a-a89c8771d593)

![instr_agtadduser](https://github.com/e107inc/e107/assets/135135923/94502fd2-4458-4424-a82d-6e3c94bd548b)

**↓ ↓ Continued in a post below ↓ ↓**

**Comment by Alex-e107nl** [comment]
How good, it looks nice, I'm curious and will test it as soon as you release it, I'm just missing a contact form per object, are you still planning that?

**Reply by Vodhin** [reply]
I've invited you to the repository. I still have a lot of work to do but it should be OK to load it up on a website that isn't important - **I would avoid installing it on a Production site.** I've done a bunch of my own _Spaghetti Code_ to allow the Plugin to make changes to Users' Admin Permissions regarding Backend Admin Access: I am fairly confident that the code is stable and should not be a security issue.

**Reply by Vodhin** [reply]
`I'm just missing a contact form per object, are you still planning that?`
Yes. A Contact Form is in place, complete with some basic JavaScript field validation. I have NOT coded the actual send functions yet, but will include integration with e107's email systems along with my own version of PM (which also will handle Saved Properties). This will need to be a little more robust since I want it to work with non-members, probably using a cookie or the browser's Data Storage.

**Comment by Vodhin** [comment]
Some Views of the Front End:

![demo001](https://github.com/e107inc/e107/assets/135135923/56af8619-881b-4ab8-a60b-997bc6eb91c1)

![demo002](https://github.com/e107inc/e107/assets/135135923/90041b40-99da-4a94-974b-db90fc0d61a7)

![demo003](https://github.com/e107inc/e107/assets/135135923/cdb9fc8a-7aed-457c-9a4f-df482735de2b)


And some better shots of the back end Edit Form:

The Listing Tab
![demo008a](https://github.com/e107inc/e107/assets/135135923/1149a99a-9bf1-408f-98d3-dc91b327b986)


Agent/Agency Selection:
![demo008b](https://github.com/e107inc/e107/assets/135135923/a025978f-c679-4a32-b225-f6e6eed11395)


In Form Agency Editing:
![demo008c](https://github.com/e107inc/e107/assets/135135923/4afefaba-d3f7-4a6f-808d-cd834a38c731)

In Form Agent Editing
![demo008d](https://github.com/e107inc/e107/assets/135135923/01586598-5f52-4601-b8a9-4d7028f7d164)

![demo008e](https://github.com/e107inc/e107/assets/135135923/ded67dce-b16a-4d6a-a6b8-4a60335bc457)


The Address Tab and In Form Select Box Editing:
![demo009a](https://github.com/e107inc/e107/assets/135135923/e8d7ff1e-9049-4fc6-a7c2-1235159056a1)

![demo009b](https://github.com/e107inc/e107/assets/135135923/2536e681-c55c-46de-8df8-fdb17aa2e8b4)

![demo009c](https://github.com/e107inc/e107/assets/135135923/2768865f-1736-40c5-ad0e-3a260499168c)


The "Spaces" (Rooms) list Tab:
Sortable Rows and Tables
![demo010a](https://github.com/e107inc/e107/assets/135135923/fb027bad-5285-464b-ac53-6782068b651b)

Editing a Space - General Information:
![demo010b](https://github.com/e107inc/e107/assets/135135923/87c4142a-0ff8-427a-b16d-63a09ccf451d)

Features for the Space:
![demo010c](https://github.com/e107inc/e107/assets/135135923/96ca049e-423b-4e51-ae23-9d9385a24276)

Uploading a new image for the Space:
![demo010d](https://github.com/e107inc/e107/assets/135135923/ebda2e77-f309-47e9-8abc-5cf10cb32bad)

Cropping an image for the space:
![demo010e](https://github.com/e107inc/e107/assets/135135923/58df5412-c159-4b92-afb0-1048e90c79c6)

**Reply by Vodhin** [reply]
Property Details Tab:
![demo011](https://github.com/e107inc/e107/assets/135135923/0dd0f0a5-3b92-4351-b40b-ae8a054abfe3)

Gallery Page (with Drag & Drop re-ordering):
![demo012a](https://github.com/e107inc/e107/assets/135135923/3a2c8e0e-03bc-48cd-bab5-a7ddb0b5758d)

In Form Image Editing:
![demo012b](https://github.com/e107inc/e107/assets/135135923/6accdeb8-e3c2-46b6-9a82-cb1ef244c3f2)

Even Calendar (needs work):
![demo013a](https://github.com/e107inc/e107/assets/135135923/907d183b-7dc0-4cbb-b104-ded8504293d4)

Adding and Event:
![demo013b](https://github.com/e107inc/e107/assets/135135923/07972407-a711-47b0-a4e7-36f88c7c82d4)

**Reply by Vodhin** [reply]
Other Forms:

The "My Profile" Form:
![demo030](https://github.com/e107inc/e107/assets/135135923/589cb220-ea13-4494-a3d5-94f4520d9a6f)

Agency Profile Form:
![demo031](https://github.com/e107inc/e107/assets/135135923/51c53a21-bc0b-4919-a0d7-607ee537fa90)

Data Preset Form:
![demo040](https://github.com/e107inc/e107/assets/135135923/2bbac189-e648-42dc-8f48-f4320eeaad02)

**Reply by Vodhin** [reply]
And the Front End Listing form (pretty much the same as the back-end but can be used for Private/Member submitted Listings):

NOTE: You can Enable/Disable the Non-Agents Listing Submissions, and if enabled Non-Agent (non-admin) users CANNOT edit Preset Data: they can only Add new Data.

![demo014](https://github.com/e107inc/e107/assets/135135923/7d918e7d-7132-4759-afbb-ded6e5a9a26a)

![demo015](https://github.com/e107inc/e107/assets/135135923/ecadb878-fbd1-407d-bcc1-77b60b3c4c3f)

![demo016](https://github.com/e107inc/e107/assets/135135923/1a4fe660-11b9-4b27-b6d2-dd42ad068f52)

![demo017](https://github.com/e107inc/e107/assets/135135923/9f6aebb0-9708-43ff-851a-b92ec4b2fd4c)

![demo018](https://github.com/e107inc/e107/assets/135135923/8eb6988e-1290-4877-8001-978340499aaa)

![demo019](https://github.com/e107inc/e107/assets/135135923/5848c896-207d-4001-85bd-010dc34c5655)

![demo020](https://github.com/e107inc/e107/assets/135135923/dfaf9b94-026e-49e7-92b5-313d2c6bcfb0)

![demo021](https://github.com/e107inc/e107/assets/135135923/f2995146-4f9d-484b-9650-b95a71018409)

![demo022](https://github.com/e107inc/e107/assets/135135923/eee99574-c80f-4bf6-a5d8-a6500abf9f2e)

**Reply by rica-carv** [reply]
> And the Front End Listing form (pretty much the same as the back-end but can be used for Private/Member submitted Listings):

Just curious, did you reuse the form code from the admin area, or you duplicated the code for the front end form?

**Reply by Vodhin** [reply]
`did you reuse the form code from the admin area..?` 
Hah! It's a mess! This is the first Plugin I have written for e107 V2x. All my other plugins were on v1.7x and earlier. That said, It was a bit of a steep learning curve to start this using the new Admin interface, and it uses a bunch of different, experimental ways to create the user interface. Currently there are 22 separate, un-associated Tables to manage the data, a 9000+ line JavaScript file (obviously not minified) to enhance the interface, and the main admin_config and e_module files jump through hoops to identify the user's access and load what they are allowed to access ( and it doesn't matter if front end or back end areas).

So, yes. The Front End forms are NOT the same coding as the Back End forms, and use different PHP scripts to produce a similar interface. Most of the JavaScript that enhances the Front End forms is copy/pasted to a separate JS file just for that form and produce the same jQuery-based Mini Forms.

There is a ton of Cleanup to simplify the interface, and with any clean up there's a ton of testing to make sure any bugs haven't snuck back in.

**Reply by kreossino** [reply]
I am available to translate the plugin into Italian. Please let me know if you would like my assistance.

**Reply by Vodhin** [reply]
Translating to Italian would be nice. I will let you know when the language files are ready - I'm hoping they will be done in early June.

**Comment by Alex-e107nl** [comment]
@Vodhin Thx! i will test it, maybe i can translate it in Dutch!

**Reply by Vodhin** [reply]
Thank you! I have many new changes that have not uploaded to the repository yet, including a lot of new Language definitions (so hold off on the translation). I'm more interested in what needs to be done to make this more internationally friendly: I am planning to enhance the` Country -> State -> County -> City/Town -> Village/Subdivision` structure: there is a Field Validation system that may be too strict.

**Comment by Vodhin** [comment]
New features added, including new Menus, Front-end Template Selection, Template Re-ordering, and a new one-way Messaging System that visitors can use to initiate contact with a Seller or Agent.


### New Sidebar Menus
![demo48](https://github.com/e107inc/e107/assets/135135923/9f831be0-73d4-4cad-a9ff-8704bc6ce286)
**The _Estate Inbox_ menu** allows Agents and Private Sellers to see messages they have received. All messages are also sent to the Agent's or Seller's email address. If the Sender is logged in, a copy of the message is also sent via the e107 PM system if it is enabled.

**The _Estate Sidebar_ menu**, if enabled,  replaces some details in a Listing's View Page Summary Section, placing the Major details, Price, and Seller's/Agent's Information here - if it displayed (the _Summary Section_ checks to see if the Sidebar Menu will be displayed in the current Theme layout).


### Dynamic Templates:
The Estate Template system now supports easy Layout Selection and Re-ordering (if the Layout supports it). This is accomplished by building upon the e107 Bootstrap Template Array() format. By extending the array to include special tags, I have been able to implement an easy to use  interface for changing the look of the Estate Plugin on your website. Re-ordering is done by simply dragging the sections up or down. You can also choose to enable/disable the sections of a re-orderable Layout.  See Below for Template Coding

**Template Selection and Layout Re-ordering** - Admin Area
![demo045](https://github.com/e107inc/e107/assets/135135923/cb352052-20a1-4933-8a2f-755d8c13b573)
![demo046](https://github.com/e107inc/e107/assets/135135923/9c777fef-af5e-4aea-b058-ee5cb8fc2122)


**Template Re-ordering** - Front End
![demo047](https://github.com/e107inc/e107/assets/135135923/2c7b1d09-f782-454d-8620-197c92e75837)
![demo048](https://github.com/e107inc/e107/assets/135135923/6637d789-58a3-479f-aaf8-48e77ae65ac7)


### The Template Coding Rules:
```
 *  Templates must use $ESTATE_TEMPLATE['area']['layout_name']['txt'] format - see below
 * 
 * Template Array: 
 *    $ESTATE_TEMPLATE['area'] = ('view','list', or 'menu');
 *    
 *  Use 'name' to Identify Layout in Estate Prefs: 
 *    $ESTATE_TEMPLATE['view']['mylayout']['name'] = 'My Layout View Template'; 
 *    
 *  Use 'txt' for the actual template, can be Fixed or array();
 *    $ESTATE_TEMPLATE['view']['mylayout']['txt'] = '<div>{SHORTCODE1}</div><div>{SHORTCODE2}</div>';
 * 
 *  View and Menu areas can be Fixed or Array
 *    Fixed Layout: 
 *    $ESTATE_TEMPLATE['view']['mylayout']['txt'] = '<div>{SHORTCODE1}</div><div>{SHORTCODE2}</div>';
 *    $ESTATE_TEMPLATE['menu']['mylayout']['txt'] = '<div>{SHORTCODE3}</div><div>{SHORTCODE4}</div>';
 *    
 *    Hard Coded Array:
 *    $ESTATE_TEMPLATE['view']['mylayout']['txt'][0] = '<div>{SHORTCODE1}</div>';
 *    $ESTATE_TEMPLATE['view']['mylayout']['txt'][1] = '<div>{SHORTCODE2}</div>';
 *    
 *    $ESTATE_TEMPLATE['menu']['mylayout']['txt'][0] = '<div>{SHORTCODE3}</div>';
 *    $ESTATE_TEMPLATE['menu']['mylayout']['txt'][1] = '<div>{SHORTCODE4}</div>';
 *    
 *    Named Array (re-orderable): 
 *    $ESTATE_TEMPLATE['view']['mylayout']['ord'] = array('section_1_name','section_2_name');
 *    $ESTATE_TEMPLATE['view']['mylayout']['txt']['section_1_name'] = '<div>{SHORTCODE1}</div>';
 *    $ESTATE_TEMPLATE['view']['mylayout']['txt']['section_2_name'] = '<div>{SHORTCODE2}</div>';
 *    
 *    $ESTATE_TEMPLATE['menu']['mylayout']['ord'] = array('menu_1_name','menu_2_name');
 *    $ESTATE_TEMPLATE['menu']['mylayout']['txt']['menu_1_name'] = '<div>{SHORTCODE3}</div>';
 *    $ESTATE_TEMPLATE['menu']['mylayout']['txt']['menu_2_name'] = '<div>{SHORTCODE4}</div>';
 *    
 *    Estate Main Admin can change order of Named Array templates via Estate Prefs and through front end 
 *    Changes affect all Listings
 *    
 *  List Template does NOT support 'txt' as an Array
 *    List Template:
 *    $ESTATE_TEMPLATE['list']['mylayout']['name'] = 'My Layout List Template'; //used for Layout Select in Estate Preferences
 *    $ESTATE_TEMPLATE['list']['mylayout']['txt'] = '<div>{SHORTCODE}</div>';
 *    foreach($data as $k=>$v){
 *      $ESTATE_TEMPLATE['list']['mylayout']['txt'] .= '<div>{SHORTCODE}</div>';
 *      } 
 *    
```



### Messaging System (and Like/Saves, too!)
The One-way Messaging system adds a simple contact Form with pre-defined message elements that a user can choose from for contacting the Agent/Seller. The message Form is manipulated by JavaScript with limited User Input needed in order to complete the form. It fills in the needed Listing Information based on the type of communication the User desires.  It also employs several BOT traps that are pretty good at fooling spamming scripts without the need for captcha (though I will be including an option to include that in the future).

These predefined messages are tailored to the Current Listing based on the Status of the Listing and Status of the Agent/Seller: An Agent can accept requests to sell a property while a Private Seller cannot. An active listing includes options to request a Viewing and Make an offer, while a Pending listing will not include requests for Viewings but will include the option to make a Backup Offer, and a Sold listing provides no contact option for (Private Sellers' Sold listings are not visible to the public). 


**Preferences in Admin Area:**
![demo044](https://github.com/e107inc/e107/assets/135135923/f8d0c26a-a069-4d11-9653-73fe0eafc183)
Set rules for how the One-way Messaging System behaves for your visitors, Sellers, and Agents.


**Sidebar  Menu** with Contact Form
![demo50](https://github.com/e107inc/e107/assets/135135923/32ef5c5e-8b68-4836-8145-1118b502461f)


**Contact Form without active Estate Sidebar Menu**
![demo51](https://github.com/e107inc/e107/assets/135135923/997285c4-3fa1-46b8-b9a3-3d6e238e7f23)



**Seller/Agent Message Inbox**
![demo52](https://github.com/e107inc/e107/assets/135135923/bf032b6a-b94e-4830-bb3b-eb1d5969b061)

**Comment by Vodhin** [comment]
### [A BETA version is now available to download](https://github.com/Vodhin/estate/releases/tag/v1.0.1)
https://github.com/Vodhin/estate/releases/tag/v1.0.1

**Reply by Vodhin** [reply]
A New Release is available, Link above has been updated.

**Comment by Vodhin** [comment]
### [A Stable Version is now available to download](https://github.com/Vodhin/estate/releases/tag/v1.0.5)
https://github.com/Vodhin/estate/releases/tag/v1.0.5 

There is still a lot of work I need to do (Help files and other little things), so keep an eye out for future updates!

**Comment by Vodhin** [comment]
[**NEW version 1.1.0 is available**](https://github.com/Vodhin/estate/releases/tag/v1.1.0)
https://github.com/Vodhin/estate/releases/tag/v1.1.0

November 27 2024: So many changes, so many fixes. Added new functions to handle PHP based Currency and Date formatting based on PHP Locale settings which can be localized for each property listing. Currency formatting can also be forced to a spcific symbol before or after the number, or omitted altogether. This is handled by a new form that replaces the old Original Price and Listing Price fields (they are still present, but hidden). Included in the new form is also a new Price History table to record changes to the listing's price or status. Added ability to save custom Listing templates for each property: these can be altered either in the Admin area or directly on a Listing's View page.
